api/grpcserver/transaction_service.go

```diff
@@ -89,11 +89,10 @@ func (s TransactionService) SubmitTransaction(ctx context.Context, in *pb.Submit
 	log.Info("GRPC TransactionService.SubmitTransaction BROADCAST tx address: %x (len: %v), amount: %v, gas limit: %v, fee: %v, id: %v, nonce: %v",
 		tx.GetRecipient(), len(tx.GetRecipient()), tx.Amount, tx.GasLimit, tx.GetFee(), tx.ID().ShortString(), tx.AccountNonce)
 
-	go func() {
-		if err := s.publisher.Publish(ctx, svm.IncomingTxProtocol, in.Transaction); err != nil {
-			log.Error("error broadcasting incoming tx: %v", err)
-		}
-	}()
+	if err := s.publisher.Publish(ctx, svm.IncomingTxProtocol, in.Transaction); err != nil {
```

Comment:
> transactions were published in separate goroutines, which sometimes resulted in an immediate failure due to the verification and sometimes out-of-order network delivery

what exactly is the verification failure? TXs should be mostly independent from each other. what kind of ordering is expected?

Reply:
projected nonce verification. for example if bot publishes 100 transactions most of them will be dropped

Reply:
they can't be totally independent. i know that we are not going to enforce strict order, but even then there will be a limit on  transactions that can be submitted concurrently.

Reply:
but it sounds more like the way bot generate txs are bot matching reality. if most TXs have different principals, then the async-ness make more sense.

Reply:
> they can't be totally independent. i know that we are not going to enforce strict order, but even then there will be a limit on transactions that can be submitted concurrently.

maybe, but the way to honor limit should be rarelimiting instead of serializing the processing.

Reply:
there is literally no reason to create a goroutine for publishing a transaction.. it doesn't enable anything but creates pretty bad artificail issues if account wants to apply transactions serially. i don't understand your point at all

Reply:
hmmm... the ratelimiting should prolly done at the RPC level.

> if account wants to apply transactions serially.

why should we even honor `apply transactions serially` at the RPC level? that should be done by arranging nonce.

i recall you changed beacon to publish proposal to a goroutine because it does synchronous self-broadcast (and cause some issue that i don't know).  in any case, i am fine with changing to sync.

Reply:
> why should we even honor apply transactions serially at the RPC level? that should be done by arranging nonce.

because then we can't add more than one transaction from the same account to the mempool reliably. so for example here is a test that sends 5 transactions from the same account https://github.com/dshulyak/systest/blob/master/tests/transactions_test.go#L74-L90 . without synchronous API i will have to wait until a transaction is added to the mempool using different means, for example listening to events. so it creates artificial usability issues.

> i recall you changed beacon to publish proposal to a goroutine because it does synchronous self-broadcast (and cause some issue that i don't know). in any case, i am fine with changing to sync.

that's a problem only if we are sending a message to ourselves and can accidentally deadlock, in beacon we are receiving a message in the same component and accidental deadlock is possible